File: packages/layout-core/src/ids.ts
/**
 * Deterministic bin codes (P3).
 *
 * Bin identity derives from structural position, not from a random UUID, so
 * re-publishing a layout never orphans inventory. The database still uses a
 * surrogate UUID primary key; `code` is the natural unique key.
 */
import type { LaneSide } from './schema.js';

export const DEFAULT_BIN_CODE_PATTERN = '{warehouse}/{aisle}/{side}/B{bay:03}/L{level}';

/** Tokens the bin code pattern may reference. Anything else is a schema error. */
export const ALLOWED_CODE_TOKENS = ['warehouse', 'aisle', 'lane', 'side', 'bay', 'level'] as const;

const TOKEN_RE = /\{(\w+)(?::0?(\d+))?\}/g;

export type BinCodeParts = {
  warehouse: string;
  aisle: string;
  lane: string;
  side: LaneSide;
  /** 1-based bay number, as humans count. */
  baySeq: number;
  /** 0-based level index; rendered with +1 so ground level reads as `L1`. */
  levelIndex: number;
};

export function sideLetter(side: LaneSide): 'L' | 'R' {
  return side === 'LEFT' ? 'L' : 'R';
}

export function tokensInPattern(pattern: string): string[] {
  return [...pattern.matchAll(TOKEN_RE)].map((m) => m[1] ?? '');
}

export function isValidCodePattern(pattern: string): boolean {
  const tokens = tokensInPattern(pattern);
  if (tokens.length === 0) return false;
  return tokens.every((t) => (ALLOWED_CODE_TOKENS as readonly string[]).includes(t));
}

/**
 * Render a bin code. Unknown tokens render as empty strings, but the schema
 * rejects unknown tokens up front so this stays a pure formatter.
 */
export function formatBinCode(pattern: string, parts: BinCodeParts, warehouseCodeFallback = 'WH'): string {
  return pattern.replace(TOKEN_RE, (_match, name: string, width?: string) => {
    let raw: string;
    switch (name) {
      case 'warehouse':
        raw = parts.warehouse || warehouseCodeFallback;
        break;
      case 'aisle':
        raw = parts.aisle;
        break;
      case 'lane':
        raw = parts.lane;
        break;
      case 'side':
        raw = sideLetter(parts.side);
        break;
      case 'bay':
        raw = String(parts.baySeq);
        break;
      case 'level':
        raw = String(parts.levelIndex + 1);
        break;
      default:
        raw = '';
    }
    return width ? raw.padStart(Number(width), '0') : raw;
  });
}

File: server/app/layout/ids.py
"""Deterministic bin codes — Python mirror of ``layout-core/src/ids.ts``."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

DEFAULT_BIN_CODE_PATTERN = "{warehouse}/{aisle}/{side}/B{bay:03}/L{level}"

ALLOWED_CODE_TOKENS = ("warehouse", "aisle", "lane", "side", "bay", "level")

_TOKEN_RE = re.compile(r"\{(\w+)(?::0?(\d+))?\}")

LaneSide = Literal["LEFT", "RIGHT"]


@dataclass(frozen=True, slots=True)
class BinCodeParts:
    warehouse: str
    aisle: str
    lane: str
    side: LaneSide
    bay_seq: int
    """1-based bay number, as humans count."""

    level_index: int
    """0-based level index; rendered with +1 so ground level reads as ``L1``."""


def side_letter(side: LaneSide) -> str:
    return "L" if side == "LEFT" else "R"


def tokens_in_pattern(pattern: str) -> list[str]:
    return [match.group(1) for match in _TOKEN_RE.finditer(pattern)]


def is_valid_code_pattern(pattern: str) -> bool:
    tokens = tokens_in_pattern(pattern)
    if not tokens:
        return False
    return all(token in ALLOWED_CODE_TOKENS for token in tokens)


def format_bin_code(pattern: str, parts: BinCodeParts) -> str:
    """Render a bin code. Unknown tokens render empty; the schema rejects them first."""

    def replace(match: re.Match[str]) -> str:
        name = match.group(1)
        width = match.group(2)
        raw = {
            "warehouse": parts.warehouse,
            "aisle": parts.aisle,
            "lane": parts.lane,
            "side": side_letter(parts.side),
            "bay": str(parts.bay_seq),
            "level": str(parts.level_index + 1),
        }.get(name, "")
        return raw.zfill(int(width)) if width else raw

    return _TOKEN_RE.sub(replace, pattern)
